nttt/acknowledgements.py

```diff
@@ -0,0 +1,74 @@
+import math
+import pandas
+import sys
+from .utilities import get_file, save_file
+
+
+def get_volunteer_acknowledgement(csv_file_path, language):
+    '''
+    Returns the volunteer acknowledgement for the given language or None if
+    the volunteer acknowledgement doesn't exist.
+    '''
+
+    data_frame = None
+    try:
+        data_frame = pandas.read_csv(csv_file_path)
+    except:
+        print("Could not read the volunteer acknowledgements spreadsheet.",
+              file=sys.stderr)
+        return None
+
+    acknowledgement = None
+    for i in data_frame.index:
+        if data_frame.at[i, 'Locale'] == language:
+            acknowledgement = data_frame.at[i, 'Acknowledgement']
+            break
+    return acknowledgement
+
+
+def add_volunteer_acknowledgement(csv_file_path, output_file_path, language,
+                                  volunteers, logging):
+    '''
+    Adds the volunteer acknowledgement to the given output file. Returns
+    whether the volunteer acknowledgement was successfully added.
+    '''
+
+    acknowledgement = get_volunteer_acknowledgement(csv_file_path, language)
+    if acknowledgement is None:
+        print("Could not find the volunteer acknowledgement for language {}".format(language),
+              file=sys.stderr)
+        return False
+    elif str(acknowledgement) == "nan":
+        print("Empty volunteer acknowledgement for language {}".format(language),
+              file=sys.stderr)
+        return False
+
+    name_placeholders = "[name]\n\n[name]\n\n[name]\n"
+    if name_placeholders not in acknowledgement:
+        print("Incorrect volunteer acknowledgement for language {}".format(language),
+              file=sys.stderr)
+        return False
+
+    if len(volunteers) == 0:
+        print("Warning: No volunteer name(s) given - please add them manually")
+    else:
+        volunteer_names = ""
+        first = True
+        for volunteer in volunteers:
+            if first:
+                first = False
+                volunteer_names = volunteer + "\n"
+            else:
+                volunteer_names = volunteer_names + "\n" + volunteer + "\n"
+        acknowledgement = acknowledgement.replace(name_placeholders, volunteer_names)
+    if logging == "on":
+        print("Volunteer acknowledgement to be added:")
+        print(acknowledgement)
+
+    content, suggested_eol = get_file(output_file_path)
+    # Strip trailings blanks (spaces, tabs, newlines) from the original content
+    # before appending the volunteer acknowledgement. This avoids a surplus of
+    # empty lines between the content and the volunteer acknowledgement.
+    content = content.rstrip() + "\n***\n" + acknowledgement + "\n"
```

Comment:
From the issue description:

> Translated volunteer acknowledgement should be added only once, when published for the first time.

It seems like the acknowledgement text is always added with this implementation.

Reply:
It looks that way, but it should not be. The first time, the acknowledgement is added to the final step file. The source of this step file is the l10n_master branch (where CrowdIn put it). The updated step file is committed to the draft branch and at some point merged to the master branch. But the updated step file is not synced to CrowdIn, so the next time the step file has to be published, it will not contain an acknowledgement, and the acknowledgement has to be added again. This is correct, isn't it @majamanojlovic?
I could add an extra check to find out whether the final step file already contains an acknowledgement, but I'm not sure if this is worth the effort.

Reply:
I think it's worth adding the check to keep nttt idempotent. Something as simple as checking for `***` is good enough.

Reply:
Agreed; added the simple check.